be/src/exec/pipeline/exchange/mem_limited_chunk_queue.h

```diff
@@ -0,0 +1,244 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mutex>
+#include <queue>
+
+#include "exec/pipeline/exchange/multi_cast_local_exchange.h"
+#include "serde/encode_context.h"
+#include "util/bit_mask.h"
+#include "util/phmap/phmap.h"
+#include "util/spinlock.h"
+
+namespace starrocks::pipeline {
+
+class MemLimitedChunkQueue {
+private:
+    struct Cell {
+        ChunkPtr chunk;
+        size_t accumulated_rows = 0;
+        size_t accumulated_bytes = 0;
+        int32_t used_count = 0;
+        bool flushed = false;
+    };
+
+    struct Block {
+        Block(int32_t consumer_number) : reader_mask(consumer_number) {}
+
+        std::vector<Cell> cells;
+        bool in_mem = true;
+        std::atomic_bool has_load_task = false;
+        // a bit mask to record which consumer is pendind read
+        // there may be a flush task between can_pop and pop,
+        // we should prevent the block that is about to be read from being flushed.
+        SpinLock lock;
+        BitMask reader_mask;
```

Comment:
using CRoaring will be ok

Reply:
> using CRoaring will be ok

we won't benefit from CRoaring in this scenario, I prefer to keep it simple as much as possible

Reply:
It is not necessary to implement an additional class with the same functionality.

Reply:
roaring::Roaring will use more memory than BitMask